### ISSUE TITLE
Group settings scanner gives violation even though all the settings match.

### ISSUE THREAD

**Issue by red2k18** [forseti-security/forseti-security#2880]
Below is the rule file config and group settings provided by the user.

rules:
  - name: All groups with iam policies should have all of these settings
    mode: whitelist
    only_iam_groups: True
    groups_emails:
      - '*'
    settings:
      #allowExternalMembers: True
      whoCanJoin: "INVITED_CAN_JOIN"
      whoCanInvite: "ALL_MANAGERS_CAN_INVITE"
      whoCanAdd: "ALL_MANAGERS_CAN_ADD"
      allowExternalMembers: False
      #whoCanLeaveGroup: "ALL_MANAGERS_CAN_LEAVE"\

**Comment by red2k18** [comment]
<?xml version="1.0" encoding="UTF-8"?>
<entry xmlns="http://www.w3.org/2005/Atom" xmlns:apps="http://schemas.google.com/apps/2006" xmlns:gd="http://schemas.google.com/g/2005">
 <id>tag:googleapis.com,2010:apps:groupssettings:GROUP:it@domain.com</id>
 <title>Groups Resource Entry</title>
 <content type="text">IT</content>
 <author>
  <name>Google</name>
 </author>
 <apps:email>it@domain.com</apps:email>
 <apps:name>IT</apps:name>
 <apps:description/>
 <apps:whoCanJoin>INVITED_CAN_JOIN</apps:whoCanJoin>
 <apps:whoCanViewMembership>ALL_MEMBERS_CAN_VIEW</apps:whoCanViewMembership>
 <apps:whoCanViewGroup>ALL_MEMBERS_CAN_VIEW</apps:whoCanViewGroup>
 <apps:whoCanInvite>ALL_MANAGERS_CAN_INVITE</apps:whoCanInvite>
 <apps:whoCanAdd>ALL_MANAGERS_CAN_ADD</apps:whoCanAdd>
 <apps:allowExternalMembers>false</apps:allowExternalMembers>
 <apps:whoCanPostMessage>ANYONE_CAN_POST</apps:whoCanPostMessage>
 <apps:allowWebPosting>true</apps:allowWebPosting>
 <apps:maxMessageBytes>26214400</apps:maxMessageBytes>
 <apps:isArchived>true</apps:isArchived>
 <apps:archiveOnly>false</apps:archiveOnly>
 <apps:messageModerationLevel>MODERATE_NONE</apps:messageModerationLevel>
 <apps:spamModerationLevel>MODERATE</apps:spamModerationLevel>
 <apps:replyTo>REPLY_TO_IGNORE</apps:replyTo>
 <apps:customReplyTo/>
 <apps:includeCustomFooter>false</apps:includeCustomFooter>
 <apps:customFooterText/>
 <apps:sendMessageDenyNotification>false</apps:sendMessageDenyNotification>
 <apps:defaultMessageDenyNotificationText/>
 <apps:showInGroupDirectory>false</apps:showInGroupDirectory>
 <apps:allowGoogleCommunication>false</apps:allowGoogleCommunication>
 <apps:membersCanPostAsTheGroup>false</apps:membersCanPostAsTheGroup>
 <apps:messageDisplayFont>DEFAULT_FONT</apps:messageDisplayFont>
 <apps:includeInGlobalAddressList>true</apps:includeInGlobalAddressList>
 <apps:whoCanLeaveGroup>ALL_MEMBERS_CAN_LEAVE</apps:whoCanLeaveGroup>
 <apps:whoCanContactOwner>ANYONE_CAN_CONTACT</apps:whoCanContactOwner>
 <apps:whoCanAddReferences>NONE</apps:whoCanAddReferences>
 <apps:whoCanAssignTopics>NONE</apps:whoCanAssignTopics>
 <apps:whoCanUnassignTopic>NONE</apps:whoCanUnassignTopic>
 <apps:whoCanTakeTopics>NONE</apps:whoCanTakeTopics>
 <apps:whoCanMarkDuplicate>NONE</apps:whoCanMarkDuplicate>
 <apps:whoCanMarkNoResponseNeeded>NONE</apps:whoCanMarkNoResponseNeeded>
 <apps:whoCanMarkFavoriteReplyOnAnyTopic>NONE</apps:whoCanMarkFavoriteReplyOnAnyTopic>
 <apps:whoCanMarkFavoriteReplyOnOwnTopic>NONE</apps:whoCanMarkFavoriteReplyOnOwnTopic>
 <apps:whoCanUnmarkFavoriteReplyOnAnyTopic>NONE</apps:whoCanUnmarkFavoriteReplyOnAnyTopic>
 <apps:whoCanEnterFreeFormTags>NONE</apps:whoCanEnterFreeFormTags>
 <apps:whoCanModifyTagsAndCategories>NONE</apps:whoCanModifyTagsAndCategories>
 <apps:favoriteRepliesOnTop>false</apps:favoriteRepliesOnTop>
 <apps:whoCanApproveMembers>ALL_MANAGERS_CAN_APPROVE</apps:whoCanApproveMembers>
 <apps:whoCanBanUsers>OWNERS_AND_MANAGERS</apps:whoCanBanUsers>
 <apps:whoCanModifyMembers>OWNERS_AND_MANAGERS</apps:whoCanModifyMembers>
 <apps:whoCanApproveMessages>OWNERS_AND_MANAGERS</apps:whoCanApproveMessages>
 <apps:whoCanDeleteAnyPost>NONE</apps:whoCanDeleteAnyPost>
 <apps:whoCanDeleteTopics>NONE</apps:whoCanDeleteTopics>
 <apps:whoCanLockTopics>NONE</apps:whoCanLockTopics>
 <apps:whoCanMoveTopicsIn>NONE</apps:whoCanMoveTopicsIn>
 <apps:whoCanMoveTopicsOut>NONE</apps:whoCanMoveTopicsOut>
 <apps:whoCanPostAnnouncements>NONE</apps:whoCanPostAnnouncements>
 <apps:whoCanHideAbuse>NONE</apps:whoCanHideAbuse>
 <apps:whoCanMakeTopicsSticky>NONE</apps:whoCanMakeTopicsSticky>
 <apps:whoCanModerateMembers>OWNERS_AND_MANAGERS</apps:whoCanModerateMembers>
 <apps:whoCanModerateContent>OWNERS_AND_MANAGERS</apps:whoCanModerateContent>
 <apps:whoCanAssistContent>NONE</apps:whoCanAssistContent>
 <apps:customRolesEnabledForSettingsToBeMerged>false</apps:customRolesEnabledForSettingsToBeMerged>
 <apps:enableCollaborativeInbox>false</apps:enableCollaborativeInbox>
 <apps:whoCanDiscoverGroup>ALL_MEMBERS_CAN_DISCOVER</apps:whoCanDiscoverGroup>

**Comment by peter-fe** [comment]
Is there any progress on the issue?

We are facing the same problem, the scanner is showing not the correct value for "allowExternalMembers".
It seems this field is always interpreted as true by the scanner.

Forseti version in use: v2.19.1

Data from the scanner:
```
{
   "allowExternalMembers":true,
   "whoCanAdd":"ALL_MANAGERS_CAN_ADD",
   "whoCanInvite":"ALL_MANAGERS_CAN_INVITE",
   "whoCanJoin":"INVITED_CAN_JOIN",
   "whoCanLeaveGroup":"ALL_MEMBERS_CAN_LEAVE",
   "whoCanViewGroup":"ALL_MEMBERS_CAN_VIEW",
   "whoCanViewMembership":"ALL_MEMBERS_CAN_VIEW"
}
```

Manual call to the Google groups API:
```
<?xml version="1.0" encoding="UTF-8"?>
<entry xmlns="http://www.w3.org/2005/Atom" xmlns:apps="http://schemas.google.com/apps/2006" xmlns:gd="http://schemas.google.com/g/2005">
 <id>tag:googleapis.com,2010:apps:groupssettings:GROUP:cscc-test@...</id>
 <title>Groups Resource Entry</title>
 <content type="text">cscc-test</content>
 <author>
  <name>Google</name>
 </author>
 <apps:email>cscc-test@...</apps:email>
 <apps:name>cscc-test</apps:name>
 <apps:description/>
 <apps:whoCanJoin>INVITED_CAN_JOIN</apps:whoCanJoin>
 <apps:whoCanViewMembership>ALL_MEMBERS_CAN_VIEW</apps:whoCanViewMembership>
 <apps:whoCanViewGroup>ALL_MEMBERS_CAN_VIEW</apps:whoCanViewGroup>
 <apps:whoCanInvite>ALL_MANAGERS_CAN_INVITE</apps:whoCanInvite>
 <apps:whoCanAdd>ALL_MANAGERS_CAN_ADD</apps:whoCanAdd>
 <apps:allowExternalMembers>false</apps:allowExternalMembers>
 <apps:whoCanPostMessage>ALL_MEMBERS_CAN_POST</apps:whoCanPostMessage>
 <apps:allowWebPosting>true</apps:allowWebPosting>
 <apps:maxMessageBytes>26214400</apps:maxMessageBytes>
 <apps:isArchived>true</apps:isArchived>
 <apps:archiveOnly>false</apps:archiveOnly>
 <apps:messageModerationLevel>MODERATE_NONE</apps:messageModerationLevel>
 <apps:spamModerationLevel>MODERATE</apps:spamModerationLevel>
 <apps:replyTo>REPLY_TO_IGNORE</apps:replyTo>
 <apps:customReplyTo/>
 <apps:includeCustomFooter>false</apps:includeCustomFooter>
 <apps:customFooterText/>
 <apps:sendMessageDenyNotification>false</apps:sendMessageDenyNotification>
 <apps:defaultMessageDenyNotificationText/>
 <apps:showInGroupDirectory>true</apps:showInGroupDirectory>
 <apps:allowGoogleCommunication>false</apps:allowGoogleCommunication>
 <apps:membersCanPostAsTheGroup>false</apps:membersCanPostAsTheGroup>
 <apps:messageDisplayFont>DEFAULT_FONT</apps:messageDisplayFont>
 <apps:includeInGlobalAddressList>true</apps:includeInGlobalAddressList>
 <apps:whoCanLeaveGroup>ALL_MEMBERS_CAN_LEAVE</apps:whoCanLeaveGroup>
 <apps:whoCanContactOwner>ANYONE_CAN_CONTACT</apps:whoCanContactOwner>
 <apps:whoCanAddReferences>NONE</apps:whoCanAddReferences>
 <apps:whoCanAssignTopics>NONE</apps:whoCanAssignTopics>
 <apps:whoCanUnassignTopic>NONE</apps:whoCanUnassignTopic>
 <apps:whoCanTakeTopics>NONE</apps:whoCanTakeTopics>
 <apps:whoCanMarkDuplicate>NONE</apps:whoCanMarkDuplicate>
 <apps:whoCanMarkNoResponseNeeded>NONE</apps:whoCanMarkNoResponseNeeded>
 <apps:whoCanMarkFavoriteReplyOnAnyTopic>NONE</apps:whoCanMarkFavoriteReplyOnAnyTopic>
 <apps:whoCanMarkFavoriteReplyOnOwnTopic>NONE</apps:whoCanMarkFavoriteReplyOnOwnTopic>
 <apps:whoCanUnmarkFavoriteReplyOnAnyTopic>NONE</apps:whoCanUnmarkFavoriteReplyOnAnyTopic>
 <apps:whoCanEnterFreeFormTags>NONE</apps:whoCanEnterFreeFormTags>
 <apps:whoCanModifyTagsAndCategories>NONE</apps:whoCanModifyTagsAndCategories>
 <apps:favoriteRepliesOnTop>true</apps:favoriteRepliesOnTop>
 <apps:whoCanApproveMembers>ALL_MANAGERS_CAN_APPROVE</apps:whoCanApproveMembers>
 <apps:whoCanBanUsers>OWNERS_AND_MANAGERS</apps:whoCanBanUsers>
 <apps:whoCanModifyMembers>OWNERS_AND_MANAGERS</apps:whoCanModifyMembers>
 <apps:whoCanApproveMessages>OWNERS_AND_MANAGERS</apps:whoCanApproveMessages>
 <apps:whoCanDeleteAnyPost>OWNERS_AND_MANAGERS</apps:whoCanDeleteAnyPost>
 <apps:whoCanDeleteTopics>OWNERS_AND_MANAGERS</apps:whoCanDeleteTopics>
 <apps:whoCanLockTopics>OWNERS_AND_MANAGERS</apps:whoCanLockTopics>
 <apps:whoCanMoveTopicsIn>OWNERS_AND_MANAGERS</apps:whoCanMoveTopicsIn>
 <apps:whoCanMoveTopicsOut>OWNERS_AND_MANAGERS</apps:whoCanMoveTopicsOut>
 <apps:whoCanPostAnnouncements>OWNERS_AND_MANAGERS</apps:whoCanPostAnnouncements>
 <apps:whoCanHideAbuse>NONE</apps:whoCanHideAbuse>
 <apps:whoCanMakeTopicsSticky>NONE</apps:whoCanMakeTopicsSticky>
 <apps:whoCanModerateMembers>OWNERS_AND_MANAGERS</apps:whoCanModerateMembers>
 <apps:whoCanModerateContent>OWNERS_AND_MANAGERS</apps:whoCanModerateContent>
 <apps:whoCanAssistContent>NONE</apps:whoCanAssistContent>
 <apps:customRolesEnabledForSettingsToBeMerged>false</apps:customRolesEnabledForSettingsToBeMerged>
 <apps:enableCollaborativeInbox>false</apps:enableCollaborativeInbox>
 <apps:whoCanDiscoverGroup>ALL_IN_DOMAIN_CAN_DISCOVER</apps:whoCanDiscoverGroup>
</entry>
```

**Comment by gkowalski-google** [comment]
Hi @peter-fe, where did you get the data from the scanner in your last comment? I have tested this with the ```allowExternalMembers: True``` commented out similar to your config and it is reporting violations. A couple of ideas to check in the `scanner` section of the Forseti config.yaml:

1. Confirm the `groups_settings` scanner is enabled
2. Confirm the `rules_path` location is correct

**Comment by peter-fe** [comment]
Hi @gkowalski-google,

Sorry for the late feedback, I was quite busy and somehow lost track about this.

To answer your questions:

1. `groups_settings` scanner is enabled
 ```
        - name: group
          enabled: true
        - name: groups_settings
          enabled: true
```
2. `rules_path` location is correct
```
    # if no rules_path is specified, rules are
    # searched in /path/to/forseti_security/rules/
    rules_path: /home/ubuntu/forseti-security/rules

ls /home/ubuntu/forseti-security/rules
audit_logging_rules.yaml            forwarding_rules.yaml                  kms_rules.yaml
bigquery_rules.yaml                 group_rules.yaml                       lien_rules.yaml
blacklist_rules.yaml                groups_settings_rules.yaml             location_rules.yaml
bucket_rules.yaml                   iam_rules.yaml                         log_sink_rules.yaml
cloudsql_rules.yaml                 iap_rules.yaml                         resource_rules.yaml
enabled_apis_rules.yaml             instance_network_interface_rules.yaml  retention_rules.yaml
external_project_access_rules.yaml  ke_rules.yaml                          role_rules.yaml
firewall_rules.yaml                 ke_scanner_rules.yaml                  service_account_key_rules.yaml
```

Regarding the scanner data, I got it directly from the cscc when looking up the findings:

![cscc](https://user-images.githubusercontent.com/22496291/65018503-9351aa00-d929-11e9-941d-907be718ade9.jpg)

Or by looking up the finding in MySQL by it's id:

```
mysql> select violation_data from violations where id = 66882;
+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| violation_data                                                                                                                                                                                                    |
+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| "rule specified (allowExternalMembers:False AND whoCanJoin:INVITED_CAN_JOIN AND whoCanInvite:ALL_MANAGERS_CAN_INVITE AND whoCanAdd:ALL_MANAGERS_CAN_ADD AND whoCanLeaveGroup:ALL_MANAGERS_CAN_LEAVE) is required" |
+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)

mysql> select resource_data from violations where id = 66882;
+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| resource_data                                                                                                                                                                                                                                                                            |
+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| {"allowExternalMembers": true, "whoCanAdd": "ALL_MANAGERS_CAN_ADD", "whoCanInvite": "ALL_MANAGERS_CAN_INVITE", "whoCanJoin": "INVITED_CAN_JOIN", "whoCanLeaveGroup": "ALL_MEMBERS_CAN_LEAVE", "whoCanViewGroup": "ALL_MEMBERS_CAN_VIEW", "whoCanViewMembership": "ALL_MEMBERS_CAN_VIEW"} |
+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```

The data from UI and DB match as expected, but is different to the actual settings of the GSuite group, as validated by the REST call and admin.google.com.

Best,
Peter

**Comment by gkowalski-google** [comment]
@peter-fe Do you have time today to look into this together? I am available from 10am - 2pm PDT. I can also meet later this afternoon.

**Comment by peter-fe** [comment]
@gkowalski-google,

Woo PDT is a tough one, 10am is about 7pm at my time zone, but I think we can get it going.
This week I'm quite busy and I've to travel, so how about next week?

**Comment by gkowalski-google** [comment]
@joecheuk Can you review this and see if you have any ideas?

**Comment by gkowalski-google** [comment]
This has been released as part of [Forseti Release v2.22.0](https://github.com/forseti-security/forseti-security/releases/tag/v2.22.0).
